cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java

```diff
@@ -122,22 +122,24 @@ void put(String key, Value value, Scope scope) {
     Preconditions.checkNotNull(value.toString(), String.format("Null value provided for the key '%s'.", key));
     Preconditions.checkState(nodeName != null, "nodeName cannot be null.");
 
-    List<NodeValue> nodeValueList = tokenValueMap.get(scope).get(key);
-    if (nodeValueList == null) {
-      nodeValueList = Lists.newArrayList();
-      tokenValueMap.get(scope).put(key, nodeValueList);
-    }
+    synchronized (tokenValueMap) {
+      List<NodeValue> nodeValueList = tokenValueMap.get(scope).get(key);
+      if (nodeValueList == null) {
+        nodeValueList = Lists.newArrayList();
+        tokenValueMap.get(scope).put(key, nodeValueList);
+      }
 
-    // Check if the current node already added the key to the token.
-    // In that case replace that entry with the new one
-    for (int i = 0; i < nodeValueList.size(); i++) {
-      if (nodeValueList.get(i).getNodeName().equals(nodeName)) {
-        nodeValueList.set(i, new NodeValue(nodeName, value));
-        return;
+      // Check if the current node already added the key to the token.
+      // In that case replace that entry with the new one
+      for (int i = 0; i < nodeValueList.size(); i++) {
+        if (nodeValueList.get(i).getNodeName().equals(nodeName)) {
+          nodeValueList.set(i, new NodeValue(nodeName, value));
+          return;
+        }
       }
-    }
 
-    nodeValueList.add(new NodeValue(nodeName, value));
+      nodeValueList.add(new NodeValue(nodeName, value));
```

Comment:
The above for loop will do replacement, but then the same value will get appended to the list as well? Meaning you'll have a duplicated value?